Add TFCLIPModel

ydshieh opened the pull request:
# What does this PR do?


Add `TFCLIPModel`, along with `TFCLIPTextModel` and `TFCLIPVisionModel`.

(and fixed doc examples in `modeling_clip.py`)

## Who can review?
@Rocketknight1 @patrickvonplaten @NielsRogge @sgugger @LysandreJik

ydshieh committed:
Start the work for TFCLIPModel

ydshieh committed:
Convert to TF code (TODO: loss + doc)

ydshieh committed:
Clean up

ydshieh committed:
Fix pooled_output for TFCLIPTextTransformer - using tf.gather_nd

ydshieh committed:
assert -> raise error

ydshieh committed:
Expose TFCLIPModel

ydshieh committed:
Deal with dummy_inputs

ydshieh committed:
Add tests

ydshieh committed:
Fix all tests. TODO: manual check weight loading + add more comments

ydshieh committed:
Fix pt tf equivalence test

ydshieh committed:
fixes

ydshieh committed:
update TFCLIPVisionEmbeddings's Conv2D

ydshieh committed:
Fix loss + overwrite test_pt_tf_model_equivalence from common

ydshieh committed:
Add a comment about the change about MainLayer in test_keras_save_load

ydshieh committed:
Set return_loss=True in TFCLIPModelTester + make tests pass

ydshieh committed:
overwrite test_pt_tf_model_equivalence from tf common

ydshieh committed:
fix base_model_prefix

ydshieh committed:
Fix examples

ydshieh committed:
remove unused

ydshieh commented:
Finally finalized this PR after working on something else. It is ready for review :)

patil-suraj commented on src/transformers/models/clip/modeling_tf_clip.py:
```diff
@@ -0,0 +1,1528 @@
+        inputs = input_processing(
+            func=self.call,
+            config=self.config,
+            input_ids=pixel_values,
```
I'm not familiar with this. But this seems nit hackish and as there are going to be more vision models IMO it should be handled explicitly rather than setting `input_ids=pixel_values`. cc @Rocketknight1

patil-suraj commented on src/transformers/models/clip/modeling_tf_clip.py:
```diff
@@ -0,0 +1,1528 @@
+        if "input_ids" in inputs:
+            inputs["pixel_values"] = inputs.pop("input_ids")
+
+        if inputs["pixel_values"] is None:
+            raise ValueError("You have to specify pixel_values")
```
We should probably check this before `input_processing`

patil-suraj commented on tests/test_modeling_tf_vit.py:
```diff
@@ -200,7 +200,7 @@ def test_model(self):
     # overwrite from common since `encoder_seq_length` and `encoder_key_length` are calculated
     # in a different way than in text models.
-    @tooslow
+    @slow
```
IMO this should be changed in a separate PR.

patil-suraj commented on src/transformers/models/clip/modeling_clip.py:
```diff
@@ -827,7 +827,7 @@ def forward(
             >>> outputs = model(**inputs)
             >>> last_hidden_state = outputs.last_hidden_state
-            >>> pooled_output = outputs.pooled_output # pooled CLS states
+            >>> pooled_output = outputs.pooler_output # pooled CLS states
```
Good catch!

patil-suraj replied:
This is fixed in #14641

patil-suraj reviewed (approved):
Thanks a lot for adding the TF version @ydshieh!

All the CLIP-related logic looks good to me! @Rocketknight1  Would be awesome if you could review the TF code :)

Rocketknight1 commented on src/transformers/activations_tf.py:
```diff
@@ -63,6 +63,12 @@ def gelu_fast(x):
     return 0.5 * x * (1.0 + tf.tanh(x * coeff2 * (1.0 + coeff1 * x * x)))
 
 
+def quick_gelu(x):
```
I really want to delete all of these approximate functions and just not support TF <= 2.3! (But let's leave that for another PR)

sgugger commented on src/transformers/models/clip/modeling_tf_clip.py:
```diff
@@ -0,0 +1,1528 @@
+        image_embeds(:obj:`tf.Tensor` of shape :obj:`(batch_size, output_dim`):
+            The image embeddings obtained by applying the projection layer to the pooled output of
+            :class:`~transformers.TFCLIPVisionModel`.
+        text_model_output(:obj:`TFBaseModelOutputWithPooling`):
```
```suggestion
        text_model_output(:class:`~transformers.modeling_tf_utils.TFBaseModelOutputWithPooling`):
```

sgugger commented on src/transformers/models/clip/modeling_tf_clip.py:
```diff
@@ -0,0 +1,1528 @@
+            :class:`~transformers.TFCLIPVisionModel`.
+        text_model_output(:obj:`TFBaseModelOutputWithPooling`):
+            The output of the :class:`~transformers.TFCLIPTextModel`.
+        vision_model_output(:obj:`TFBaseModelOutputWithPooling`):
```
```suggestion
        vision_model_output(:class:`~transformers.modeling_tf_utils.TFBaseModelOutputWithPooling`):
```

sgugger commented on src/transformers/models/clip/modeling_tf_clip.py:
```diff
@@ -0,0 +1,1528 @@
+
+        batch_size, num_channels, height, width = shape_list(pixel_values)
+
+        # When running on CPU, `tf.nn.conv2d` doesn't support `NCHW` format.
+        # So change the input format from `NCHW` to `NHWC`.
+        # shape = (batch_size, in_height, in_width, in_channels=num_channels)
+        pixel_values = tf.transpose(pixel_values, perm=(0, 2, 3, 1))
```
Should we properly document we expect inputs in the NCHW format? I think TF users might be more used to NHWC?

Rocketknight1 replied:
@ydshieh I discussed this with Sylvain, and we decided keeping NCHW is definitely better!

ydshieh replied:
`CLIP_VISION_INPUTS_DOCSTRING` has already mentioned

```
pixel_values: ... and each example must have the shape :obj:`(batch_size, num_channels, height, width)`):
```
I just added (for `TFCLIPVisionEmbeddings`)
```
  def call(self, pixel_values: tf.Tensor) -> tf.Tensor:
      """`pixel_values` is expected to be of NCHW format."""
```

sgugger commented on src/transformers/models/clip/modeling_tf_clip.py:
```diff
@@ -0,0 +1,1528 @@
+
+    def build(self, input_shape: tf.TensorShape):
+
+        # Q: should we use normal initializer?
+        # initializer = tf.random_normal_initializer(mean=0.0, stddev=1.0)
```
Needs to be addressed @patil-suraj do you have an answer?

patil-suraj replied:
Aah yes! CLIP's initialization logic is bit complex, indeed we should use `normal` here.
See the ` CLIPPreTrainedModel.__init_weights` method for reference.

https://github.com/huggingface/transformers/blob/4cdb67caba9abee8b5b94b4f88f55e95a3d1013f/src/transformers/models/clip/modeling_clip.py#L345

sgugger commented on src/transformers/models/clip/modeling_tf_clip.py:
```diff
@@ -0,0 +1,1528 @@
+        attention_output = tf.reshape(tensor=attention_output, shape=(batch_size, -1, self.embed_dim))
+
+        attention_output = self.out_proj(attention_output, training=training)
+        # In TFBert, attention weights are returned after dropout.
+        # However, in CLIP, they are returned before dropout.
+        # Q: What to do?
```
This should be consistent with the PyTorch version.

Rocketknight1 replied:
+1 on this - match the PyTorch CLIP implementation!

sgugger commented on src/transformers/models/clip/modeling_tf_clip.py:
```diff
@@ -0,0 +1,1528 @@
+    ) -> Union[TFBaseModelOutputWithPooling, Tuple[tf.Tensor]]:
+        input_shape = shape_list(input_ids)
+
+        embedding_output = self.embeddings(
+            input_ids=input_ids,
+            position_ids=position_ids,
+        )
```
```suggestion
        embedding_output = self.embeddings(input_ids=input_ids, position_ids=position_ids)
```
Fits in one line.

sgugger commented on src/transformers/models/clip/modeling_tf_clip.py:
```diff
@@ -0,0 +1,1528 @@
+class TFCLIPTextMainLayer(tf.keras.layers.Layer):
+    config_class = CLIPTextConfig
+
+    def __init__(self, config: CLIPTextConfig, **kwargs):
+        super().__init__(**kwargs)
+
+        self.config = config
+
+        self.text_model = TFCLIPTextTransformer(config, name="text_model")
```
No need for all those new lines IMO.

sgugger commented on src/transformers/models/clip/modeling_tf_clip.py:
```diff
@@ -0,0 +1,1528 @@
+    config_class = CLIPVisionConfig
+
+    def __init__(self, config: CLIPVisionConfig, **kwargs):
+        super().__init__(**kwargs)
+
+        self.config = config
+
+        self.vision_model = TFCLIPVisionTransformer(config, name="vision_model")
```
Same here: no need for all the new lines.

sgugger commented on src/transformers/models/clip/modeling_tf_clip.py:
```diff
@@ -0,0 +1,1528 @@
+        VISION_DUMMY_INPUTS = tf.random.uniform(
+            shape=(len(DUMMY_INPUTS), 3, self.config.image_size, self.config.image_size), dtype=tf.float32
+        )
+        return {
+            "pixel_values": VISION_DUMMY_INPUTS,
+        }
```
```suggestion
        return {"pixel_values": VISION_DUMMY_INPUTS}
```

sgugger commented on tests/test_modeling_tf_common.py:
```diff
@@ -276,12 +276,19 @@ def test_onnx_runtime_optimize(self):
             for module_member_name in dir(module)
             if module_member_name.endswith("MainLayer")
+            # This condition is required, since `modeling_tf_clip.py` has 3 classes whose names end with `MainLayer`.
+            and module_member_name[: -len("MainLayer")] == model_class.__name__[: -len("Model")]
```
Can we group the two together with a single
```
if module_member_name.replace("MainLayer", "Model") == model_class.__name__
```

ydshieh replied:
if `module_member_name` itself is something like `TFCLIP(Text/Vision)Model`, the condition in your comment will hold, but we only want to deal with `TFCLIP(Text/Vision)MainLayer`, right?

sgugger commented on tests/test_modeling_tf_common.py:
```diff
@@ -276,12 +276,19 @@ def test_onnx_runtime_optimize(self):
     def test_keras_save_load(self):
         config, inputs_dict = self.model_tester.prepare_config_and_inputs_for_common()
 
+        # `CLIPModelTester/TFCLIPModelTester` set `return_loss` to `True`, however this causes the preparation
+        # of `symbolic_inputs` below failed.
+        if self.__class__.__name__ == "TFCLIPModelTest":
+            inputs_dict.pop("return_loss", None)
+
```
This should not be in the common tests, but the clip tester should overwrite this test.

ydshieh replied:
OK

sgugger reviewed (approved):
Thanks a lot for adding this model!

Rocketknight1 commented on src/transformers/models/clip/modeling_tf_clip.py:
```diff
@@ -0,0 +1,1528 @@
+        )
+
+
+class TFCLIPPreTrainedModel(TFPreTrainedModel):
```
The PyTorch version of this class also included [the weight initializations for all of the layers in the model](https://github.com/huggingface/transformers/blob/master/src/transformers/models/clip/modeling_clip.py#L345-L386) - I agree that this probably isn't the right place to put them in TF, but we should make sure that we use the same initialization values.

ydshieh commented on src/transformers/models/clip/modeling_tf_clip.py:
```diff
@@ -0,0 +1,1528 @@
+            position_ids = tf.expand_dims(tf.range(start=0, limit=input_shape[-1]), axis=0)
+
+        position_embeds = tf.gather(params=self.position_embedding, indices=position_ids)
+        position_embeds = tf.tile(input=position_embeds, multiples=(input_shape[0], 1, 1))
```
@Rocketknight1 Good point. Yes, it is possible. However

- The line `position_embeds = tf.tile(input=position_embeds, multiples=(input_shape[0], 1, 1))` is everywhere in TF model scripts. If we change here, probably it is better to change all (?)
- `tf.broadcast_to` is not perfect as mentioned in TF doc [broadcast_to](https://www.tensorflow.org/api_docs/python/tf/broadcast_to)

  ```
  When doing broadcasted operations such as multiplying a tensor by a scalar,
  broadcasting (usually) confers some time or space benefit, as the broadcasted
  tensor is never materialized.
  
  However, broadcast_to does not carry with it any such benefits. The newly-created
  tensor takes the full memory of the broadcasted shape. (In a graph context, broadcast_to
  might be fused to subsequent operation and then be optimized away, however.)
  ```

But after diving deeper, I found something more severe: 

The `...INPUTS_DOCSTRING` (everywhere in transformers) specifies `position_ids` should have shape `(batch_size, sequence_length)`. If a user specify `position_ids` manually with the full `batch_size` (very unlikely though), then the line `tf.tile` shouldn't be run! (otherwise the `batch_size` will become squared). 

If we use `tf.broadcast_to`, we don't have such issue.

Could you discuss with @sgugger  and @LysandreJik  to see what to do about this, please?

Rocketknight1 replied:
I don't think we need `tf.broadcast_to` either - because the first dimension of `position_embeds` is 1, we can just remove the tile line entirely, and position_embeds will still be correctly broadcast in the next line. In other words, the code would just be:
```
position_embeds = tf.gather(params=self.position_embedding, indices=position_ids)
final_embeddings = inputs_embeds + position_embeds
```

I did some quick testing and this seems to work fine. However, it's just a nit, and the performance impact will likely be small anyway, so don't worry about it too much! We can also just leave it, and I'll make a note later to check the codebase for instances of `tile` that should be replaced with implicit broadcasting.

Rocketknight1 replied:
Also, thinking about it more, using broadcasting resolves your issue with the users specifying a full batch of `position_ids`, so maybe it's more than just a minor performance nit!

ydshieh replied:
> 

Thanks, @Rocketknight1 

Well, implicit broadcasting will still work when users specifying a full batch - there won't be any broadcasting in this case.

But as you said, we can leave it to another PR if necessary :)

Rocketknight1 reviewed (approved):
Overall, this is a great PR! I have a couple of nits in addition to the comments raised by other people, but the main one is that I think layer initializations don't actually match the ones from the PyTorch model, because I don't think the values from the `_init_weights` function in `CLIPPreTrainedModel` are present anywhere in the TF file. Obviously, this isn't a problem for pretrained models, but it's definitely worth fixing for anyone who wants to train their own model from scratch.

ydshieh reviewed (commented):
I just saw that all my previous own reviews are in pending status.
(I was thinking why I didn't have any reply ...)

I have answered some questions and raised some questions too.
I tag some of you in each review comments, hope not too spamming.

ydshieh committed:
Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

ydshieh committed:
apply review suggestions

ydshieh committed:
change self.pre_layrnorm to self.pre_layernorm

ydshieh committed:
apply more review suggestions

ydshieh commented on src/transformers/models/clip/modeling_tf_clip.py:
```diff
@@ -0,0 +1,1528 @@
+        # # expand attention_mask
+        # if attention_mask is not None:
+        #     # [bsz, seq_len] -> [bsz, 1, tgt_seq_len, src_seq_len]
+        #     attention_mask = _expand_mask(attention_mask, hidden_states.dtype)
```
I was mixing/using another way to expand the mask. I changed it back to use `_expand_mask` which aligns with `CLIPTextTransformer`

ydshieh committed:
return attention probs before dropout (to align with PT)

ydshieh committed:
fix weight init

ydshieh commented:
I addressed all the reviews, in particular the weight initialization.

In PT version, there is 

```
if isinstance(module, CLIPTextEmbeddings):
    module.token_embedding.weight.data.normal_(mean=0.0, std=factor * 0.02)
    module.position_embedding.weight.data.normal_(mean=0.0, std=factor * 0.02)
```

I am not sure why `0.02` is used instead of `config.initializer_range`. cc @patil-suraj

ydshieh committed:
fix

LysandreJik commented:
Should this be merged @Rocketknight1 @patil-suraj? :smiley:

ydshieh commented:
> Should this be merged @Rocketknight1 @patil-suraj? 😃

There is (new) conflicts I just saw. I will resolve later today or tomorrow.

patil-suraj commented:
okay for merge on my side (once the conflicts are resolved) !

sgugger commented:
Merging will require:
- a rebase on master
- converting all docstrings to Markdown.

You can install our `doc-builder` utility to do this:
```
pip install git+https://github.com/huggingface/doc-builder
```

Then run from your branch in the Transformers repo
```
doc-builder convert src/transformers/models/clip/modeling_tf_clip.py
```
to do this in the newly added modeling_tf_clip file.

Note that clip.rst has disappeared and you will need to add your model to clip.mdx instead.

Let us know if you need any help and please ping for review when you're done!

ydshieh committed:
build doc

ydshieh committed:
fix missing doc

ydshieh committed:
fix for test

ydshieh commented:
@patil-suraj @Rocketknight1 @sgugger I rebased on master and fixed errors. The last failed test is irrelevant I think.
Ready to be merged when you have some time :)

Thank you for reviewing this PR!

Rocketknight1 commented:
@sgugger I'm happy at this point!

sgugger reviewed (approved):
Thanks a lot for making all those changes!

Failures are unrelated, so merging.
